src/Workspaces/CSharp/Portable/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs

```diff
@@ -1857,22 +1857,28 @@ private IEnumerable<TypeInferenceInfo> InferTypeForReturnStatement(ReturnStateme
                     return;
                 }
 
-                var ancestorExpressions = returnStatement.GetAncestorsOrThis<ExpressionSyntax>();
+                var ancestor = returnStatement.AncestorsAndSelf().FirstOrDefault(e => e is ParenthesizedLambdaExpressionSyntax || e is SimpleLambdaExpressionSyntax ||
+                    e is AnonymousMethodExpressionSyntax || e is LocalFunctionStatementSyntax);
```

Comment:
you can simplify this to `e is AnonymousFunctionSyntax || e is LocalFunctionStatementSyntax`

Reply:
`ParenthesizedLambdaExpressionSyntax` and `SimpleLambdaExpressionSyntax` derive from `LambdaExpressionSyntax` which together with `AnonymousMethodExpressionSyntax` derive from `AnonymousFunctionSyntax `

Reply:
Good one, I have updated it.